File: leecode/code/139/139.java
class Solution {
    public boolean wordBreak(String s, List<String> wordDict) {
        Map<String, Integer> map = new HashMap();
        for (int i = 0; i < wordDict.size(); i++) {
            map.put(wordDict.get(i), map.getOrDefault(wordDict.get(i), 0) + 1);
        }
        boolean[] dp = new boolean[s.length() + 1];
        dp[0] = true;

        for (int i = 1; i <= s.length(); i++) {
            for (int j = 0; j < i; j++) {
                if (dp[j] && map.containsKey(s.substring(j, i))) {
                    dp[i] = true;
                }
            }
        }
        return dp[s.length()];
    }
}

File: leecode/code/139/139.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        dp = [False] * len(s)
        dp.insert(0, True)

        for i in range(1, len(s) + 1):
            for j in range(0, i):
                tmp = s[j: i]
                if dp[j] and tmp in wordDict:
                    dp[i] = dp[j]


        return dp[len(s)]
